app/controllers/reactions_controller.rb

```diff
@@ -0,0 +1,29 @@
+class ReactionsController < ApplicationController
+  before_action :require_authentication
+  before_action :find_or_initialize_reaction, only: %i[show update]
+
+  def show
```

Comment:
I may have missed this – is this route being used anywhere?

Reply:
Yes - all links clicked in emails are get requests. I couldn't find a way around it.

Reply:
Ahh yes. We _may_ want to separate out the email reaction link into a separate action and add a signature / token to the link (like in the unsubscribe link), for a couple of reasons:

1. Otherwise it's possible for someone to craft a URL and have it perform the action on behalf of a user (by getting the user to click on it) – this is a form of [CSRF](https://www.owasp.org/index.php/Cross-Site_Request_Forgery_(CSRF)) although probably a minor security concern for us.
2. This `GET` link _may_ be triggered automatically by robots / search engines – of course, the way you've set this up it'll give an error (as the request won't be authenticated), but I'm not sure how that will affect SEO, etc.

Given this, do you reckon it's worth changing the way this is done? (I'm 50/50, for what it's worth).

Reply:
Hey - I found some bugs and have fixed the way this works a little bit.

Do you want to take another look and let me know if I've changed your mind at all!

Reply:
It feels a tad messy and overloaded. Suggestion:

- Don't reuse the `show` action for this – get rid of it, and…
- … instead, have a new action `GET set_from_email` (more explicit on what it's doing).
- … which you can now use in the email when setting a **new** reaction.
- Then in the email, don't show anything if the user already has a reaction for that lesson (my thinking here is: it's a low probability that this will happen). Or just have a link that says something like "you've already reacted to this lesson – open the lesson page to see more". This avoids having a `GET` endpoint that performs a destructive update or delete.
- In the `find_or_initialize_reaction` filter, only call `SignedGlobalID.find` if `action.to_sym == :set_from_email`, just to make sure the intention is more explicit and it only gets called then.

Also note that when attempting to find the user via the signed token we should error if no user is found, like in the Subscriptions controller: https://github.com/chaynHQ/soulmedicine/blob/88491d9ffb728a77b37b40e5abf9f7728e5b15a3/app/controllers/subscriptions_controller.rb#L63

(Give me a shout on Slack if you have any questions/thoughts on this, or if I can help in any way).

Reply:
Have updated again - totally agree with your thoughts. This is 100% tidied, what do you think?

Reply:
Looks great!